### PR TITLE
Fix: Replacing List instances with Position

### PR DESCRIPTION
**Pull request by ahmedheakl**
Fix - Position Class

## 📜 Summary
As it is inconveniece to have positions stored as `Tuple[Union[int, float], ...]` and their methods as thrown all over the place, it was only sane to move all of the data and their corresponding methods to single class under the name `Position`. 

###### Ticket(s):None

- 🔄 **Main changes in this PR**
    - Created `Position` class.
    - Added generic typing for loading Position with `float` for agent steps and `int` for difficulty calculations. 
    - Replaced List instances in `utils` and `envs` modules with `Position`.

- ⚠️ **Breaking changes**
  - Now to access the robot position -> `robot.pos.x`.
  - All methods in `teacher_checker.py` and `agent.py` modules accept only `Position` objects. 

## 🚀 How to test this Pull Request

 - Run `pytest test`

## 📝 Notes for future improvements

- Optimize `teacher_checker.py` module.

### CODE REVIEW

File: src/highrl/utils/abstract.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        """Cross product with current point and other point"""
+        return self.cross(self, other)
+
+    @staticmethod
+    def cross(current: TPosition, other: TPosition) -> T:
+        """Cross products between two positions"""
+        return np.cross(current.get_coords(), other.get_coords()).item()
+
+    def triangle_cross(self, left_pos: TPosition, right_pos: TPosition) -> T:
+        """Triangular cross product between current points and two other points"""
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/utils/abstract.py:50]: current point not points

File: src/highrl/agents/agent.py [hunk 2]
```diff
@@ -193,55 +168,56 @@ def compute_position(self, action: Any, delta_t: float) -> Tuple:
...
-        self.check_validity(action)
-        pos = self.compute_position(action, delta_t)
-        self.px, self.py, self.theta = pos
-        self.vx = action.vx
-        self.vy = action.vy
-        self.w = action.w
-
-    def fix(self, x, mod):
+    def fix(self, base: Union[int, float], mod: Union[int, float]) -> Union[int, float]:
         """Fix input x to be in range [0:mod-1].
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/agents/agent.py:180]: Not clear what does the `fix` function do exactly

File: src/highrl/agents/agent.py [hunk 3]
```diff
@@ -193,55 +168,56 @@ def compute_position(self, action: Any, delta_t: float) -> Tuple:
...
+        self.check_validity(action)
+        pos = self.compute_position(action, delta_t)
+        x_pos, y_pos, self.theta = pos
+        self.pos.set_pos(x_pos, y_pos)
+        self.vx = action.vx
+        self.vy = action.vy
+        self.w = action.w
 
     def is_overlapped(self, obstacle: SingleObstacle, check_target: str = "agent"):
         """Checks if overlap between the agent/goal and an obstacle.
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/agents/agent.py:223]: Replace doc-string with `"""Checks if there is overlap between the agent/goal and a given obstacle"""`

File: src/highrl/utils/teacher_checker.py [hunk 4]
```diff
@@ -1,13 +1,9 @@
...
+from typing import Tuple, List, Dict
 import numpy as np
 
 from highrl.obstacle.obstacles import Obstacles
-from highrl.utils.calculations import (
-    cross_product_point_line,
-    cross_product_triangle,
-    point_to_point_distance,
-)
+from highrl.utils.abstract import Position
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/utils/teacher_checker.py:6]: Replace with 
`from highrl.utils import Position`

File: src/highrl/utils/calculations.py [hunk 5]
```diff
@@ -79,78 +56,13 @@ def point_to_obstacle_distance(
...
     """Exponentiation for negative numbers
 
-    Args:
-        base (float): Base of exponent
-        exponent (float): Exponent
-
-    Returns:
-        float: sign * (|base|**exponont)
+    Since exponentiation of negative numbers might result
+    in complex numbers, we omit this by returning sign * |base| ** exponent
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/utils/calculations.py:63]: Why did you delete the following part of the doc-string?

  Args:
        base (float): Base of exponent
        exponent (float): Exponent
    Returns:
        float: sign * (|base|**exponont)

File: test/test_agent.py [hunk 6]
```diff
@@ -104,27 +119,33 @@ def test_agent_negative_x_y_velocity(self):
             expected[-1], value[-1], msg=f"Expected: {expected[-1]}, Found: {value[-1]}"
         )
 
-    def test_reached_destination_true(self):
-        agent = Agent(gx=20, gy=3)
+    def test_reached_destination_true(self) -> None:
+        """Testing whether robot reached the destination"""
+        gpos = Position[float](x_pos=20.0, y_pos=3.0)
+        agent = Agent(gpos)
```
Comment (AbdelrahmanAbdelgwad) [inline test/test_agent.py:125]: how is this not reversed with `test_reached_destination_false`

File: test/test_obstacles.py [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
 from highrl.obstacle.single_obstacle import SingleObstacle
-import numpy as np
 
 
 class ObstaclesTest(unittest.TestCase):
-    def test_flatten_contours_single_point(self):
+    """Testing obstacles creations and manipulations"""
+
+    def test_flatten_contours_single_point(self) -> None:
+        """Testing flattening the contours"""
```
Comment (AbdelrahmanAbdelgwad) [inline test/test_obstacles.py:13]: Write doc-string for the function `test_flatten_contours_single_point`

File: test/test_obstacles.py [hunk 8]
```diff
@@ -17,7 +22,8 @@ def test_flatten_contours_single_point(self):
             msg=f"Expected: {expected}, Found: {value}",
         )
 
-    def test_flatten_contours_multiple_points(self):
+    def test_flatten_contours_multiple_points(self) -> None:
+        """Testing flattening the contours"""
```
Comment (AbdelrahmanAbdelgwad) [inline test/test_obstacles.py:26]: Same here

File: test/test_teacherchecker.py [hunk 9]
```diff
@@ -63,42 +59,54 @@ def test_check_if_point_inside(self):
...
-            obstacles, coords, 6, 8, [2, 3], [4, 7], omit_first_four=False
+            obstacles, coords, 6, 8, pos, gpos, omit_first_four=False
         )
         expected = False
         self.assertEqual(
             expected, value, msg=f"\nExpected: {expected} || Found: {value}"
         )
 
     def test_convex_polygen_compute_easy(self):
+        """Testing convex hull main problem"""
```
Comment (AbdelrahmanAbdelgwad) [inline test/test_teacherchecker.py:89]: Please elaborate further in doc-string

File: src/highrl/utils/calculations.py (hunk 5 above)
Comment (ahmedheakl) [reply src/highrl/utils/calculations.py:63]: It is redundant, as these details can already be inferred from the type hints and the new doc-string.

**Review by AbdelrahmanAbdelgwad — CHANGES_REQUESTED** [review]
LGTM!! 
Just modify the few notes I have written and we are good to go

**Review by AbdelrahmanAbdelgwad — APPROVED** [review]
LGTM!!
Great Job
